File: User_Friendly_Interpolation_Search/Interpolation_Search.cpp
#include <iostream>

using namespace std;

int interpolation(int arr[], int N, int x)
{
    int flag = 0, pos, low = 0, high = N - 1, mid;

    if(x > arr[high] || x < arr[low])
        return -1;

    while(low < high)
    {
        mid = low + ((high - low) * (x - arr[low]) / (arr[high] - arr[low]));
        
        if(arr[mid] == x)
        {
            flag = 1;
            pos = mid + 1;
            return pos;
        }
        
        if(x > arr[mid])
            low = mid + 1;
        else
            high = mid - 1;
    }
    
    return -1;
}

int main()
{
    int x;
    int arr[100];
    int N, pos;

    cout << "Enter the number of elements : ";
    cin >> N;
    cout << "Enter the elements in sorted order : ";
    
    for(int i = 0; i < N; i++)
        cin >> arr[i];
    
    cout << "Enter the element to be searched : ";
    cin >> x;
    
    pos = interpolation(arr, N, x);
    
    if(pos != -1)
        cout << "Element found at position " << pos << endl;
    else
        cout << "Element not found" << endl;

    return 0;
}

File: Searching/Interpolation_Search/Interpolation_Search.cs
using System;

namespace Interpolation_Search
{
    class Program
    {
        public static int interpolation(int[] a, int n, int search_item) // Function implementing Interpolation_Search
        {
            int high = n - 1;
            int low = 0;
            int pos;

            while(low <= high && search_item >= a[low] && search_item <= a[high])
            {
                int rise = high - low;
                int run = a[high] - a[low];
                int x = search_item - a[low];
                pos = low + (rise / run) * x;

                if(a[pos] == search_item)
                    return pos;
                else if(search_item < a[pos])
                    high = pos - 1;
                else if(search_item > a[pos])
                    low = pos + 1;
            }
            return -1;
        }

        public static void Main(String[] args)
        {
            int[] a = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10}; // Sorted Input array
            int n = a.GetLength(0); // Size of array
            int search_item = 8; // Element to be searched
            int index = interpolation(a, n, search_item);

            if(index == -1)
                Console.WriteLine("Element not found");
            else
                Console.WriteLine("Found at position " + index);
            Console.WriteLine();
            Console.ReadLine();
        }
    }
}

/* Output

Found at position 7

*/
